planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+    }
+
+The custom stream and event types could be used to define segregations between different types of event logs within a single application as part
+of the partitioning/sharding/segmenting strategy
+
+
+
+The schema generation would conform to the shape of the base event and stream wrapper much the way that document storage does in the Doc Db features.
+
+### Table/Function Generation and Stream Identity
+
+TODO -- allow more flexibility in the identity of streams, so `int`, `long`, `Guid`, `string` just like the document types. Do more code generation
+to accomodate this.
+
+TODO -- there will be more columns & even new tables to track and coordinate the asynchronous projections
+
+### Event Store Partitioning or Sharding
+
+The event sourcing in Marten has limited scalability because of its dependence today upon appending events to 
+one table. Closely related to the storage is ways to segment the asynchronous projection support to parallelize
+the work across multiple threads or active nodes.
+
+Here are some ideas:
+
+* Allow the user to define multiple event logs in the database. This way you'd have multiple stream/event tables for completely
+  different kinds of streams
+* Postgresql native partitioning by tenant id where that's useful
+* Aggregation keys -- this is tightly coupled to the projections. If there's a projection that is done by something other than tenant id or stream id, try to persist the 
+  key for however the events/streams are grouped to stream and/or event storage. Thinking about data elements like "Region" or "Country" where you might be doing
+  aggregations across streams or tenants. Maybe that's useful?
+* Archive streams that reach some kind of termination event? Move everything to an archive table off to the side? Keeps the event log table cleaner
+* Artificial segmenting of the streams. Use some kind of algorithm that could see stream ids, and assign them to 2 or more "cohorts", then persist that "cohort key" to 
+  both the stream and event data. This would potentially help for the async daemon to parallelize work. It would definitely help out for "live" projections by having 
+  smaller tables to work through
+
+### Projections
+
+Thoughts:
+
+Functionality wise, the lifecycle of projections are:
+
+* "Live" projections built on demand from the raw events. This only changes for v4 to take advantage of new snapshotting capabilities
+* "Inline" projections built as part of the same unit of work as capturing the source events. See the section on *Inline Projections* because this
+  will change internally for v4
+* "Async" projections built up in a background process of some sort. I think this turns into a near re-write for v4
+
+By *type*, the projection types *could* be:
+
+* Aggregate document by stream (existing)
+* Aggregate document by tenant id (new)
+* Execute SQL statement by event type (new) -- so, insert/update/delete. This comes up a lot, and it should be low hanging fruit
```

Comment:
Also using patch api?

Reply:
Yeah, I included that in the transformer idea for document based projections.